Dev

luciafalcinelli opened the pull request.

luciafalcinelli committed:
added the permissions for the users

luciafalcinelli committed:
added invitation users

luciafalcinelli committed:
Merge branch 'dev' into admin_users

luciafalcinelli committed:
Merge pull request #10 from stats4sd/admin_users

Admin users

luciafalcinelli committed:
fixed the problem with output

luciafalcinelli committed:
deleted the pipfile

luciafalcinelli committed:
added the link to the websito on the invitation email

dave-mills commented on app/Http/Controllers/Admin/BeneficiaryTypeCrudController.php:
```diff
@@ -40,12 +40,20 @@ public function setup()
+        $this->crud->addFilter([ 
+            'type'  => 'simple',
+            'name'  => 'is_other',
+            'label' => 'Show no other',
```
Not sure that “show no other” as a label for a filter will make sense to people, as I expect it won’t be clear what “is_other” means. Maybe something like “show default list”.

luciafalcinelli replied:
Yeah, maybe you are right. the 'other' work comes from Other on choices

dave-mills commented on app/Http/Controllers/Admin/ActionCrudController.php:
```diff
@@ -49,17 +54,90 @@ public function setup()
+
+                        $query->where('output_id', $value);
+                    });
+
```
Looking at the Action model, there is actually both a `subactivities()` and an `activities()` relationship - powered by seperate link tables in the database. So this can actually be simplified - see my upcoming commit.

luciafalcinelli replied:
I don't sure how I can simplify the filter

dave-mills replied:
I've already done that - as I said, see my added commits 👍

dave-mills commented on app/Http/Controllers/Admin/ActionCrudController.php:
```diff
@@ -49,17 +54,90 @@ public function setup()
+                foreach (json_decode($values) as $key => $value) {
+                $this->crud->query = $this->crud->query->whereHas('subactivities', function ($query) use ($value) {
+                    $query->where('activity_id', $value);
+                });
```
Again here - this can be simplified to check the action <--> activity relationship directly.

dave-mills committed:
adds output_id getter to actions via activities:
 - assumes that ALL activities attached to an action are from the same output. (Seems reasonable, given you can only select 1 output when editing an action)
 - updates activity and output filters to rely on the action->activities() relationship, not the action->subactivities() relationship.

dave-mills committed:
updates effect filters to use action->activities() where relevant

dave-mills committed:
removes reference to depreciated json_decode() function:
 - standard PHP json_decode() works just as well and isn't going to vanish within the next few updates

dave-mills reviewed (commented):
In testing, I found that when editing an action, the output field was not reset correctly - it was always set to Output 1 (the default option) regardless of what I had chosen for the action previously.

Looks like this is because the output isn't saved on the action (which is correct, as it's just a way of filtering the activities). So I added a getter to the Action model, that calculates the output_id based on the activities. 

Doing this, I also noticed that action->activities() is a valid relationship as well as action->subactivities(). I do now remember a conversation about this - this is to allow for actions that do not have subactivities. So I tweaked the output and activity filters to use this activities() relationship, which will avoid issues when actions don't have subactivities. 

I also have a question about who receives the invite email - see my comment in the inviteCrudController. 

I think once you've checked and confirmed who should receive the new user login details, this is good to go   👍

dave-mills commented on app/Http/Controllers/Admin/InvitationCrudController.php:
```diff
@@ -0,0 +1,130 @@
+        $user->save();
+        $user->teams()->sync($request['teams']); 
+        
+        Mail::to(backpack_user())->send(new InviteUser($user, $text_password));
```
The email here is currently being sent to the user who creates the invite. Is this correct? Seems counter-intuitive to me, as I would assume the new user should receive their new login details directly. But this could be desired functionality - maybe the admin user should get the new user's login details and then pass them on manually? 

Let me know if this is intended here :)

luciafalcinelli replied:
No, it is wrong. Sorry my bad

dave-mills committed:
adds files related to my-account ready to customise

dave-mills committed:
adds automatic message to new users to change their passwords:
 - all users will receive a messge on the main dashboard, prompting them to change their passwords.
 - Once they do, the pw_changed flag is set in the database and they no longer see the prompt.
 - This is a soft prompt, as users are not prevented from using the platform before changing their password

dave-mills committed:
updates invitation email with clearer text:
 - sets up  as option to use when sending email from the platform
 - adds instructions to new users about changing their password on first login.

dave-mills commented on app/Mail/InviteUser.php:
```diff
@@ -0,0 +1,42 @@
+     */
+    public function build()
+    {
+        return $this->from('lucia@stats4sd.org')
```
I feel automated invitation emails shouldn't come from an individual - it might be fine right now with at most a few users, but if this scales, I doubt you'd want dozens or hundreds of people all emailing you directly when they forget their passwords... 

I've setup a config('mail.admin_email') entry, which uses the ADMIN_EMAIL_ADDRESS env variable, or a default of support@stats4sd.org. If you want to keep the invites going from your personal email for now while you provide close support for the first few people using the platform, then leave this for now. But I suggest that at some point you swap this out for the config('mail.admin_email') so that we can both access any support requests that come from the platform.

luciafalcinelli replied:
I think using the support@stats4sd.org is the best idea. I used mine, just to test the emails.

dave-mills committed:
Fixes problem with down() method

Forgot to update the down() method to do the opposite of the up() method here...

luciafalcinelli committed:
set up the sending email to the user from ADMIN_EMAIL_ADDRESS and added the default value for pw_changed in the users

luciafalcinelli committed:
changed the env() to config() in InviteUser
